### PR TITLE
Xiaoxu/spectral dogfood

### PR DESCRIPTION
**Pull request by xiaoxuqi-ms**

### CODE REVIEW

File: packages/rulesets/src/spectral/test/default-in-enum.test.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+beforeAll(async () => {
+  linter = await linterForRule('az-default-in-enum');
+  return linter;
+});
+
+test('az-default-in-enum should find errors', () => {
+  const myOpenApiDocument = {
+    swagger: '2.0',
+    paths: {
+      definitions: {
```
Comment (jianyexi) [inline packages/rulesets/src/spectral/test/default-in-enum.test.ts:15]: it seems not a valid swagger

File: packages/rulesets/src/spectral/test/default-in-enum.test.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            "Bar"
+          ],
+          default: "Baz"
+        }
+      }
+    },
+  };
+  return linter.run(myOpenApiDocument).then((results) => {
+    expect(results.length).toBe(1);
+    expect(results[0].path.join('.')).toBe('paths.definitions.Test');
```
Comment (jianyexi) [inline packages/rulesets/src/spectral/test/default-in-enum.test.ts:33]: could you check the rule message as well

File: packages/rulesets/src/spectral/test/default-in-enum.test.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            "Bar"
+          ],
+          default: "Baz"
+        }
+      }
+    },
+  };
+  return linter.run(myOpenApiDocument).then((results) => {
+    expect(results.length).toBe(2);
+    expect(results[0].path.join('.')).toBe('paths.definitions.Test0');
```
Comment (jianyexi) [inline packages/rulesets/src/spectral/test/default-in-enum.test.ts:63]: check rule message

File: packages/rulesets/src/spectral/az-common.ts [hunk 4]
```diff
@@ -424,6 +427,37 @@ const ruleset : any = {
...
+        "function": enumInsteadOfBoolean
+      }
+    },
+    "az-avoid-anonymous-parameter": {
+      "description": "Inline/anonymous models must not be used, instead define a schema with a model name in the \"definitions\" section and refer to it. This allows operations to share the models.",
+      "message": "Inline/anonymous models must not be used, instead define a schema with a model name in the \"definitions\" section and refer to it. This allows operations to share the models.",
+      "severity": "error",
+      "resolved": false,
+      "formats": [oas2],
+      "given": ["$..parameters.*", "$.paths[*].parameters", "$.paths.*[get,put,post,patch,delete,options,head].parameters"],
```
Comment (jianyexi) [inline packages/rulesets/src/spectral/az-common.ts:457]: not sure if the first given condition is necessary?

File: packages/rulesets/src/spectral/az-common.ts [hunk 5]
```diff
@@ -424,6 +427,37 @@ const ruleset : any = {
...
+        "function": defaultInEnum
+      }
+    },
+    "az-enum-insteadOf-boolean": {
+      "description": "Booleans properties are not descriptive in all cases and can make them to use, evaluate whether is makes sense to keep the property as boolean or turn it into an enum.",
+      "message": "Booleans properties are not descriptive in all cases and can make them to use, evaluate whether is makes sense to keep the property as boolean or turn it into an enum.",
+      "severity": "warn",
+      "formats": [oas2],
+      "given": "$..[?(@.type === 'boolean')]",
+      "then": {
```
Comment (jianyexi) [inline packages/rulesets/src/spectral/az-common.ts:447]: should be apply to an unresolved swagger

Comment (jianyexi) [inline packages/rulesets/src/spectral/az-common.ts:447]: should be apply to an unresolved swagger

File: packages/rulesets/src/spectral/functions/avoid-anonymous-parameter.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+const avoidAnonymousParameter = (parameters:any, _opts:any, paths:any) => {
+  if (parameters === null || parameters.schema === undefined || parameters["x-ms-client-name"] !== undefined) {
+    return [];
+  }
+  const path = paths.path || paths.target || [];
+
+  const properties: object = parameters.schema.properties;
+  if ((properties === undefined || Object.keys(properties).length === 0) &&
+      parameters.additionalProperties === undefined &&
+      parameters.allOf === undefined) {
```
Comment (jianyexi) [inline packages/rulesets/src/spectral/functions/avoid-anonymous-parameter.ts:12]: should be parameters.schema.additionalProperties  and parametes.schema.allOf ?

Comment (jianyexi) [inline packages/rulesets/src/spectral/functions/avoid-anonymous-parameter.ts:12]: should be parameters.schema.additionalProperties  and parametes.schema.allOf ?

File: packages/rulesets/src/spectral/az-common.ts [hunk 7]
```diff
@@ -424,6 +427,38 @@ const ruleset : any = {
       "then": {
         "function": versionpolicy
       }
+    },
+    "az-default-in-enum": {
+      "description": "This rule applies when the value specified by the default property does not appear in the enum constraint for a schema.",
+      "message": "Default value should appear in the enum constraint for a schema",
+      "severity": "error",
+      "formats": [oas2],
+      "given": "$..[?(@.enum)]",
```
Comment (jianyexi) [inline packages/rulesets/src/spectral/az-common.ts:436]: this rule should apply to unresolved swagger too

File: packages/rulesets/src/spectral/functions/default-in-enum.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+const defaultInEnum = (swaggerObj:any, _opts:any, paths:any) => {
+  const defaultValue = swaggerObj.default;
+  const enumValue: string[] = swaggerObj.enum;
+  if (swaggerObj === null || typeof swaggerObj !== 'object') {
+    return [];
+  }
+  const path = paths.path || paths.target || [];
+
+  if (enumValue && !enumValue.includes(defaultValue)) {
+    return [{
```
Comment (jianyexi) [inline packages/rulesets/src/spectral/functions/default-in-enum.ts:12]: what if defaultValue is not defined ?

File: packages/rulesets/src/spectral/az-common.ts (hunk 7 above)
Comment (jianyexi) [reply packages/rulesets/src/spectral/az-common.ts:436]: the matched enum can be a property name , like 
foo: {
   properties: {
     "enum": {
        type: string
        enum : ["1", "2"]
     }
 }
}

 should use more strict condition to match  to match the second enum

**Review by jianyexi — CHANGES_REQUESTED** [review]

File: packages/rulesets/src/spectral/az-common.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (xiaoxuqi-ms) [reply packages/rulesets/src/spectral/az-common.ts]: No, it should only apply to operation parameters. I update the condition but it seems github did not receive my update. Let me check.

File: packages/rulesets/src/spectral/functions/default-in-enum.ts (hunk 8 above)
Comment (xiaoxuqi-ms) [reply packages/rulesets/src/spectral/functions/default-in-enum.ts:12]: It should also be reported as a violation.

Comment (xiaoxuqi-ms) [reply packages/rulesets/src/spectral/functions/default-in-enum.ts:12]: I checked donet code, if either default value or enum is undefined, it is valid .

File: packages/rulesets/src/spectral/az-common.ts (hunk 7 above)
Comment (xiaoxuqi-ms) [reply packages/rulesets/src/spectral/az-common.ts:437]: I will check the type inside of function "defaultInEnum"

**Review by jianyexi — APPROVED** [review]

**Pull request merged**